geomesa-memory/src/main/java/org/locationtech/geomesa/memory/cqengine/index/GeoIndex.java

```diff
@@ -0,0 +1,194 @@
+                index.remove(env, object);
+                modified = true;
+            }
+
+            return modified;
+        }
+        finally {
+            objectSet.close();
+        }
+    }
+
+    @Override
+    public void clear(QueryOptions queryOptions) {
+        this.index = new BucketIndex<SimpleFeature>(360, 180, new Envelope(-180.0, 180.0, -90.0, 90.0));
+    }
+
+    @Override
+    public ResultSet<O> retrieve(final Query<O> query, final QueryOptions queryOptions) {
+        return new ResultSet<O>() {
+            @Override
+            public Iterator<O> iterator() {
+                scala.collection.Iterator<SimpleFeature> iter = getSimpleFeatureIteratorInternal((Intersects) query, queryOptions);
+
+                // JNH: Fix this:!?
+                return (Iterator<O>) JavaConversions.asJavaIterator(iter);
+            }
+
+            @Override
+            public boolean contains(O object) {
+                final Intersects intersects = (Intersects) query;
+                String attributeName = intersects.getAttributeName();
+                Geometry geom = (Geometry) object.getAttribute(attributeName);
+                return intersects.matchesValue(geom, queryOptions);
+            }
+
+            @Override
+            public boolean matches(O object) {
+                return query.matches(object, queryOptions);
+            }
+
+            @Override
+            public Query<O> getQuery() {
+                return query;
+            }
+
+            @Override
+            public QueryOptions getQueryOptions() {
+                return queryOptions;
+            }
+
+            @Override
+            public int getRetrievalCost() {
+                return INDEX_RETRIEVAL_COST;
+            }
+
+            // JNH: I'm returning the size here as the MergeCost.
+            // The geoindex size isn't optimal, so I don't know if this is good or not...
+            @Override
+            public int getMergeCost() {
+                //System.out.println("In retrieve's getMergeCost method");
+                return size();
+            }
+
+            @Override
+            public int size() {
+                //System.out.println("In retrieve's size method");
+                return getSimpleFeatureIteratorInternal((Intersects) query, queryOptions).size();
+            }
+
+            @Override
+            public void close() {
+
+            }
+        };
+    }
+
+    private scala.collection.Iterator<SimpleFeature> getSimpleFeatureIteratorInternal(Intersects query, final QueryOptions queryOptions) {
+        final Intersects intersects = query;
+        Envelope queryEnvelope = intersects.getEnvelope();
+        return index.query(queryEnvelope, new AbstractFunction1<SimpleFeature, Object>() {
+            @Override
+            public Object apply(SimpleFeature feature) {
+                String attributeName = intersects.getAttributeName();
```

Comment:
this could be moved outside the function - we could also look up the attribute by index instead of name with a little more work